agent/src/main/java/io/pyroscope/javaagent/impl/ContinuousProfilingScheduler.java

```diff
@@ -38,14 +39,76 @@ public ContinuousProfilingScheduler(Config config, Exporter exporter, Logger log
 
     @Override
     public void start(Profiler profiler) {
-        Duration firstProfilingDuration;
+        this.logger.log(Logger.Level.DEBUG, "ContinuousProfilingScheduler#start");
+        synchronized (lock) {
+            if (started) {
+                throw new IllegalStateException("already started");
+            }
+            Duration firstProfilingDuration;
+            try {
+                firstProfilingDuration = startFirst(profiler);
+            } catch (Throwable throwable) {
+                stopSchedulerLocked();
+                throw new IllegalStateException(throwable);
+            }
+            this.profiler = profiler;
+            this.executor = Executors.newSingleThreadScheduledExecutor(THREAD_FACTORY);
+            this.job = executor.scheduleAtFixedRate(this::schedulerTick,
+                firstProfilingDuration.toMillis(), config.uploadInterval.toMillis(), TimeUnit.MILLISECONDS);
+            this.started = true;
+        }
+    }
+
+    @Override
+    public void stop() {
+        ScheduledExecutorService svc = this.executor;
+        synchronized (lock) {
+            stopSchedulerLocked();
+            this.executor = null;
+        }
+        if (svc != null) {
+            awaitTermination(svc);
+        }
```

Comment:
Do we want this inside the synchronized block? Otherwise we can have another executor active while the old one is shutting down (which is maybe fine?)

Reply:
I've left the comment in the code, and copypaste it here as well

```
        // shutdown here not under lock to avoid deadlock ( the task may block to wait for lock and
        // we are holding the lock and waiting for task to finish)
        // There is still synchronization happens from the PyroscopeAgent class,
        // so there are no concurrent calls to start/stop. So there is no lock here
```

Does it make sense?

Reply:
You are right, the synchronization in PyroscopeAgent will prevent the scenario I described. Someone could potentially misuse this class in the future, but it is unlikely and the impact is low.